### PR TITLE
new corner refinement method :: using the contour-lines

### PR DESCRIPTION
**Pull request by szk1509**
### This pullrequest changes:

The pullrequest is trying to do an extension of the "_doCornerRefinement_" parameter.
The idea is to use the contour-points to calculate the corners, in some cases more accurately.

#### Steps (roughly):

1. Marker was detected.
2. Group and classify the contour-points according to their position.
3. Evaluates the direction it follows (clockwise or not). This is for saving the corners at the right position.
4. Fit a line who passes through the points (for each group).
5. Find the Points where the edge lines crosses.
6. Save the updated corners.

### CODE REVIEW

File: modules/aruco/src/aruco.cpp [hunk 1]
```diff
@@ -771,13 +784,153 @@ class MarkerSubpixelParallel : public ParallelLoopBody {
...
+}
 
+/**
+ * Refine Corners using the contour vector :: Called from function detectMarkers
+ * @param nContours, contour-container
+ * @param nCorners, candidate Corners
+ * @param camMatrix, cameraMatrix input 3x3 floating-point camera matrix
+ * @param distCoeff, distCoeffs vector of distortion coefficient
+ */
+static void refineCandidateLines(std::vector<Point>& nContours, std::vector<Point2f>& nCorners, const Mat& camMatrix, const Mat& distCoeff){
```
Comment (szk1509) [inline modules/aruco/src/aruco.cpp]: done, i hope i didn't forget anything :)
thank you so much for your help

File: modules/aruco/src/aruco.cpp [hunk 2]
```diff
@@ -771,13 +784,153 @@ class MarkerSubpixelParallel : public ParallelLoopBody {
...
+/**
+ * Find the Point where the lines crosses :: Called from function refineCandidateLines
+ * @param nLine1
+ * @param nLine2
+ * @return Crossed Point
+ */
+static Point2f _getCrossPoint(Point3f nLine1, Point3f nLine2){
+	Matx22f A(nLine1.x, nLine1.y, nLine2.x, nLine2.y);
+	Vec2f B(-nLine1.z, -nLine2.z);
+	return Vec2f(A.solve(B).val);
```
Comment (paroj) [inline modules/aruco/src/aruco.cpp]: last warning is here, but does not come from this change but from `determinant()` usage in `Matx_FastSolveOp`

Comment (paroj) [inline modules/aruco/src/aruco.cpp]: see: https://github.com/opencv/opencv/pull/8206

**Comment by paroj** [comment]
@alalek can you trigger a rebuild for this one as the required changes are in opencv now?

File: modules/aruco/src/aruco.cpp [hunk 3]
```diff
@@ -819,6 +972,17 @@ void detectMarkers(InputArray _image, const Ptr<Dictionary> &_dictionary, Output
         parallel_for_(Range(0, _corners.cols()),
                       MarkerSubpixelParallel(&grey, _corners, _params));
     }
+
+    /// STEP 4, Optional : Corner refinement :: use contour container
+    if( _params->cornerRefinementMethod == CORNER_REFINE_CONTOUR){
+
+        if(! _ids.empty()){
+            _refineCandidateLines(contours[0], candidates[0], camMatrix.getMat(), distCoeff.getMat());
```
Comment (sovrasov) [inline modules/aruco/src/aruco.cpp]: Why do we have refinement only for `candidates[0]` here?

Comment (szk1509) [inline modules/aruco/src/aruco.cpp]: D: sorry... i was a little bit confused... my bad
Thank you all! :)

**Comment by sovrasov** [comment]
:+1:

**Comment by szk1509** [comment]
Hi @alalek :)
After the Thumbs up I've changed a tiny little comment on the code and i regret it, because it is taking so long to get into the queue (and with it the pull request) so i was wondering... should i just undo the last commit? or maybe should i just keep waiting (because it is normal that it takes so long)

sorry about bothering you (this is my first time) and thank you so much for your help :)

**Comment by alalek** [comment]
@szk1509 
This patch is not queued automatically because the source branch has reserved name "master" (this should be avoided to eliminate conflicts with other OpenCV repositories: opencv/opencv_extra).
Please select appropriate name in the next time. Thank you for contribution!

**Pull request merged**